ksql-rest-client/src/test/java/io/confluent/ksql/rest/client/FakeApiServer.java

```diff
@@ -0,0 +1,165 @@
+  private volatile int port;
+
+  private HttpMethod httpMethod;
+  private String path;
+  private Buffer body;
+  private MultiMap headers;
+
+  private Object responseObject;
+  private Buffer responseBuffer;
+  private volatile CompletableFuture<Buffer> bodyFuture;
+  private boolean connectionClosed;
+  private int errorCode = -1;
+
+  public FakeApiServer(final HttpServerOptions httpServerOptions) {
+    this.httpServerOptions = httpServerOptions;
+  }
+
+  @Override
+  public void start(final Promise<Void> startPromise) {
+    httpServer = vertx.createHttpServer(httpServerOptions).requestHandler(setupRouter())
+        .exceptionHandler(FakeApiServer::unhandledExceptionHandler);
+    httpServer.listen(ar -> {
+      if (ar.succeeded()) {
+        port = ar.result().actualPort();
+        startPromise.complete();
+      } else {
+        startPromise.fail(ar.cause());
+      }
+    });
+  }
+
+  @Override
+  public void stop(final Promise<Void> stopPromise) {
+    if (httpServer == null) {
+      stopPromise.complete();
+    } else {
+      httpServer.close(stopPromise.future());
+    }
+  }
+
+  private Router setupRouter() {
+    final Router router = Router.router(vertx);
+    router.route()
+        .handler(BodyHandler.create())
+        .handler(this::handleRequest);
+    return router;
+  }
+
+  private static void unhandledExceptionHandler(Throwable t) {
+    t.printStackTrace();
+  }
+
+  private synchronized void handleRequest(final RoutingContext routingContext) {
+    HttpServerRequest request = routingContext.request();
+    request.connection().closeHandler(v -> connectionClosed());
+
+    httpMethod = request.method();
+    path = request.path();
+    headers = request.headers();
+    body = routingContext.getBody();
+    if (bodyFuture != null) {
+      bodyFuture.complete(body);
+    }
+    if (errorCode != -1) {
+      request.response().setStatusCode(errorCode);
+    }
+    if (responseBuffer != null) {
+      request.response().end(responseBuffer);
+    } else if (responseObject != null) {
+      request.response().end(KsqlClientUtil.serialize(responseObject));
+    } else {
+      request.response().end();
+    }
+  }
+
+  public synchronized HttpMethod getHttpMethod() {
+    return httpMethod;
+  }
+
+  public synchronized String getPath() {
+    return path;
+  }
+
+  public Buffer waitForRequestBody() throws Exception {
+    this.bodyFuture = new CompletableFuture<>();
+    return bodyFuture.get();
+  }
+
+  public synchronized Buffer getBody() {
+    return body;
+  }
+
+  public synchronized MultiMap getHeaders() {
+    return headers;
+  }
+
+  public synchronized void setResponseObject(final Object responseBody) {
+    this.responseObject = responseBody;
+  }
+
+  public synchronized void setResponseBuffer(final Buffer responseBuffer) {
+    this.responseBuffer = responseBuffer;
+  }
+
+  public int getPort() {
+    return port;
+  }
+
+  private synchronized void connectionClosed() {
```

Comment:
nit: rename to `closeConnection()` or something similar?

Reply:
This is a handler that is called after the connection is closed, so I think it is ok.